cmd/tempo-cli/cmd-query-trace-summary.go

```diff
@@ -33,27 +36,96 @@ func (cmd *queryTraceSummaryCmd) Run(ctx *globalOptions) error {
 		return err
 	}
 
-	results, err := queryBucket(context.Background(), r, c, cmd.TenantID, id)
+	traceSummary, err := queryBucketForSummary(context.Background(), r, c, cmd.TenantID, id)
 	if err != nil {
 		return err
 	}
 
 	var (
-		combiner   = trace.NewCombiner()
 		marshaller = new(jsonpb.Marshaler)
 		jsonBytes  = bytes.Buffer{}
 	)
 
-	fmt.Println()
-	for i, result := range results {
-		combiner.ConsumeWithFinal(result.trace, i == len(results)-1)
+	// jsonify rootspan
+	err = marshaller.Marshal(&jsonBytes, traceSummary.RootSpan)
+	if err != nil {
+		fmt.Println("failed to marshal to json: ", err)
+		return nil
+	}
+
+	fmt.Printf("Number of blocks: %d \n", traceSummary.NumBlock)
+	fmt.Printf("Span count: %d \n", traceSummary.SpanCount)
+	fmt.Printf("Trace size: %d MB \n", traceSummary.TraceSize)
+	fmt.Printf("Trace duration: %d seconds \n", traceSummary.TraceDuration)
+	fmt.Printf("Root service name: %s \n", traceSummary.RootServiceName)
+	fmt.Println("Root span info:")
+	fmt.Println(jsonBytes.String())
+	fmt.Println("top frequent service.names: ")
+	fmt.Println(traceSummary.ServiceNames)
+
+	return nil
+}
+
+type TraceSummary struct {
+	NumBlock         int
+	SpanCount        int
+	TraceSize        int
+	TraceDuration    uint64
+	RootServiceName  string
+	RootSpan         *v1.Span
+	RootSpanResource *v1resource.Resource
+	ServiceNames     []string
+}
+
+func sortServiceNames(nameFrequencies map[string]int) PairList {
+	pl := make(PairList, len(nameFrequencies))
+	i := 0
+	for k, v := range nameFrequencies {
+		pl[i] = Pair{k, v}
+		i++
+	}
+	sort.Sort(sort.Reverse(pl))
+	return pl
+}
+
+func queryBucketForSummary(ctx context.Context, r backend.Reader, c backend.Compactor, tenantID string, traceID common.ID) (*TraceSummary, error) {
+	blockIDs, err := r.Blocks(context.Background(), tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("total blocks to search: ", len(blockIDs))
+
+	// Load in parallel
+	wg := boundedwaitgroup.New(100)
+	resultsCh := make(chan queryResults, len(blockIDs))
```

Comment:
probably better to do an unbuffered (or buffered 1) and drain the channel in a different goroutine.